### ISSUE TITLE
TimescaleDB fill rate, storage requirements, and retention policey

### ISSUE THREAD

**Issue by olizilla** [filecoin-project/sentinel#137]
The current spacerace prod db for sentinel used ~15% of it's 512GB disk in the last 7 days. That's a fill rate of 76GB per week or ~10GB per day.

The db has been migrated to a 1024GB disk and it's at 47% full. At 76GB per week, we'll need to migrate to a larger disk again in roughly 6 weeks. Let's say before, the 26th of October 2020.

#### TODO

- [ ] Research and enable compression in the db - https://docs.timescale.com/latest/using-timescaledb/compression
- [ ] Migrate to a much larger disk (we dont want to have to upgrade every 6 weeks)
- [ ] Define a retention policy

#### Notes

> It can be moved out [archieved] and deleted. Its large because we store the same data over and over again, the message cid, peerid, and hostname could be moved out of the table and converted to a foreign key which would reduce the usage dramatically i'm pretty sure. – @travisperson

**Comment by placer14** [comment]
Reading through the link, it appears compression is applied via a transform of table chunks to columular structure which can be queried (if the struct is indexed properly) but cannot be INSERTED, UPDATED, or DELETED into (yet). This makes compression directly dependent on our retention policy.

As Travis's note at the bottom of the OP mentions, there's no need to retain redundant data and could probably improve our base schema design as a first step. The plan I'm thinking of for Sentinel data flow looks something like this:

1. Data is pushed into a write-optimized schema (fast network and disk io for high bandwidth transfers, no index, no views, no relationships, denormalized as-need).
This is what Sentinel is doing today, albiet with room for improvement on drone-sourced tables where there is tag-set redundancy. A quick solution might be setting `tags_as_foreignkeys = true` on postgres output plugin which should normalize the tags automatically into a foreign-keyed table.

2. Visor runs ETL jobs (extract, transform, load) from the write-optimized schema into query-optimized schema (indexed, normalized for queries from Grafana, doesn't need high bandwidth on public interfaces--only to write-optimized schema for fast transfer). This schema should be logically separate to allow it to live physically independent of the first should the need arise. This is where we would apply the remaining normalization. These ETL jobs would run often such that the above disk usage concern is minimized. This is also where I think we apply our retention policy which would likely introduce some third stage of processing.

3. *Timescale automatically manages archive/compression/trim of data according to our compression/retention policies.* Likely pointing at some schema which is aggregate-friendly, less resolution yet "sufficient" for robust analytics in the future, and kept in a glacial type storage for infrequent reading only.

Once Visor is delivered, we can look at implementing the improvements toward improving what we do in step 1 today and enabling step 2. All the while, collecting our intuition around what "sufficient" means for step 3 once we're ready to act on it. It's probably worth representing "improvement of step 1 and clear steps toward step 2" as its own issue which preceeds this.

FYI: @frrist @iand

**Comment by iand** [comment]
Enabling compression on Timescale hypertables could reduce storage requirements dramatically and improve performance, but we can only do this once we are sure backfill is complete. Compression also prevents future schema changes. However it is possible to compress specific chunks so potentially we could identify chunks that are complete.

Partitioning the database is also an option, although this limits what kind of joins we can perform.  Timescale 2.0 offers a clustering capability that automatically queries across multiple database nodes which would allow us to scale each node's storage independently.

At the current rate of growth we have about 9 months before disk constraints require us to compress or migrate data to alternate database layouts. We should continue to investigate compression first and a plan for partitioning in the future.
